// @aeternity/aepp-sdk/es/utils/crypto.mjs
import _buffer from "buffer";
const {
  Buffer: _Buffer
} = _buffer;
import nacl from 'tweetnacl';
// js extension is required for mjs build, not importing the whole package to reduce bundle size
// eslint-disable-next-line import/extensions
import _blakejsBlake2bJs from 'blakejs/blake2b.js';
const {
  blake2b
} = _blakejsBlake2bJs;
import _varuintBitcoin from 'varuint-bitcoin';
const {
  encode: varuintEncode
} = _varuintBitcoin;
import { concatBuffers } from "./other.mjs";
import { decode, encode, Encoding } from "./encoder.mjs";
import { ArgumentError } from "./errors.mjs";
/**
 * Generate address from secret key
 * @param secret - Private key as hex string
 * @returns Public key encoded as address
 */
export function getAddressFromPriv(secret) {
  const secretBuffer = typeof secret === 'string' ? _Buffer.from(secret, 'hex') : secret;
  const keys = nacl.sign.keyPair.fromSecretKey(secretBuffer);
  return encode(keys.publicKey, Encoding.AccountAddress);
}

/**
 * Check if address is valid
 * @param address - Address
 * @param prefix - Transaction prefix. Default: 'ak'
 * @returns is valid
 */
export function isAddressValid(address) {
  let prefix = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : Encoding.AccountAddress;
  try {
    decode(address);
    const actualPrefix = address.split('_')[0];
    if (actualPrefix !== prefix) {
      throw new ArgumentError('Encoded string type', prefix, actualPrefix);
    }
    return true;
  } catch (e) {
    return false;
  }
}

/**
 * Generate a random salt (positive integer)
 * @returns random salt
 */
export function genSalt() {
  const [random] = new BigUint64Array(nacl.randomBytes(8).buffer);
  return Number(random % BigInt(Number.MAX_SAFE_INTEGER));
}

/**
 * Converts a positive integer to the smallest possible
 * representation in a binary digit representation
 * @param value - Value to encode
 * @returns Encoded number
 */
export function encodeUnsigned(value) {
  const binary = _Buffer.allocUnsafe(4);
  binary.writeUInt32BE(value);
  return binary.slice(binary.findIndex(i => i !== 0));
}

/**
 * Calculate 256bits Blake2b hash of `input`
 * @param input - Data to hash
 * @returns Hash
 */
export function hash(input) {
  return _Buffer.from(blake2b(input, undefined, 32)); // 256 bits
}

// Todo Duplicated in tx builder. remove
/**
 * Compute contract address
 * @category contract
 * @param owner - Address of contract owner
 * @param nonce - Round when contract was created
 * @returns Contract address
 */
export function encodeContractAddress(owner, nonce) {
  const publicKey = decode(owner);
  const binary = concatBuffers([publicKey, encodeUnsigned(nonce)]);
  return encode(hash(binary), Encoding.ContractAddress);
}

// KEY-PAIR HELPERS

/**
 * Generate keyPair from secret key
 * @param secret - secret key
 * @returns Object with Private(privateKey) and Public(publicKey) keys
 */
export function generateKeyPairFromSecret(secret) {
  return nacl.sign.keyPair.fromSecretKey(secret);
}

/**
 * Generate a random ED25519 keypair
 * @param raw - Whether to return raw (binary) keys
 * @returns Key pair
 */

export function generateKeyPair() {
  let raw = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
  const keyPair = nacl.sign.keyPair();
  const publicBuffer = _Buffer.from(keyPair.publicKey);
  const secretBuffer = _Buffer.from(keyPair.secretKey);
  if (raw) {
    return {
      publicKey: publicBuffer,
      secretKey: secretBuffer
    };
  }
  return {
    publicKey: encode(publicBuffer, Encoding.AccountAddress),
    secretKey: secretBuffer.toString('hex')
  };
}

// SIGNATURES

/**
 * Generate signature
 * @param data - Data to sign
 * @param privateKey - Key to sign with
 * @returns Signature
 */
export function sign(data, privateKey) {
  return nacl.sign.detached(_Buffer.from(data), _Buffer.from(privateKey));
}

/**
 * Verify that signature was signed by public key
 * @param data - Data that was signed
 * @param signature - Signature of data
 * @param address - Address to verify against
 * @returns is data was signed by address
 */
export function verify(data, signature, address) {
  return nacl.sign.detached.verify(data, signature, decode(address));
}
const messagePrefix = _Buffer.from('aeternity Signed Message:\n', 'utf8');
export const messagePrefixLength = varuintEncode(messagePrefix.length);

// TODO: consider rename to hashMessage
export function messageToHash(message) {
  const msg = _Buffer.from(message, 'utf8');
  return hash(concatBuffers([messagePrefixLength, messagePrefix, varuintEncode(msg.length), msg]));
}
export function signMessage(message, privateKey) {
  return sign(messageToHash(message), privateKey);
}

/**
 * Verify that message was signed by address
 * @param message - Message that was signed
 * @param signature - Signature of message
 * @param address - Address to verify against
 * @returns is data was signed by address
 */
// TODO: deprecate in favour of `verify(messageToHash(message), ...`, also the name is confusing
// it should contain "signature"
export function verifyMessage(message, signature, address) {
  return verify(messageToHash(message), signature, address);
}

/**
 * Check key pair for validity
 *
 * Signs a message, and then verifies that signature
 * @param privateKey - Private key to verify
 * @param publicKey - Public key to verify as hex string
 * @returns Valid?
 */
export function isValidKeypair(privateKey, publicKey) {
  const message = _Buffer.from('TheMessage');
  const signature = sign(message, privateKey);
  const publicKeyBuffer = typeof publicKey === 'string' ? _Buffer.from(publicKey, 'hex') : publicKey;
  return verify(message, signature, encode(publicKeyBuffer, Encoding.AccountAddress));
}
