// @aeternity/aepp-sdk/es/deprecated.mjs
import _defineProperty from "@babel/runtime-corejs3/helpers/defineProperty";
import _aeternityAeppCalldata from '@aeternity/aepp-calldata';
const {
  TypeResolver,
  ContractByteArrayEncoder
} = _aeternityAeppCalldata;
import { RpcError } from "./aepp-wallet-communication/schema.mjs";
/**
 * @category exception
 * @deprecated this exception is not thrown anymore
 */
export class RpcBroadcastError extends RpcError {
  constructor(data) {
    super('Broadcast failed');
    _defineProperty(this, "code", 3);
    this.data = data;
    this.name = 'RpcBroadcastError';
  }
}

/**
 * @deprecated use isAuctionName instead
 */
_defineProperty(RpcBroadcastError, "code", 3);
export const NAME_BID_MAX_LENGTH = 12; // # this is the max length for a domain to be part of a bid

/**
 * @deprecated use ContractByteArrayEncoder:encodeWithType from \@aeternity/aepp-calldata
 */
export function encodeFateValue(value, aci) {
  const valueType = new TypeResolver().resolveType(aci, {});
  return new ContractByteArrayEncoder().encodeWithType(value, valueType);
}

/**
 * @deprecated use ContractByteArrayEncoder:decodeWithType from \@aeternity/aepp-calldata
 */
export function decodeFateValue(value, aci) {
  const valueType = new TypeResolver().resolveType(aci, {});
  return new ContractByteArrayEncoder().decodeWithType(value, valueType);
}
